### ISSUE TITLE
Tesseract OCR Addon Performance

### ISSUE THREAD

**Issue by shiribailem** [friendica/friendica#14048]
I've recently upgraded and turned on the new Tesseract addon. I know it's early and due for a lot of overhaul and changes, but submitting this more for tracking sake rather than complaint.

There seems to be very little resource management with using Tesseract, it quickly used up all memory on my system repeatedly resulting in the OOM Killer firing off.

On top of that, the way it's laid out I'm not even sure what it's doing?

I mentioned in my previous suggestion of this Addon ( #13061 ) that there's a javascript implementation which I recommend investigating to offload the usage to more as-needed and client side. Proactively running Tesseract on everything uses *alot* of resources.

**Comment by shiribailem** [comment]
For reference, javascript port of the same tool: https://tesseract.projectnaptha.com/

**Comment by MrPetovan** [comment]
> alot

😏

https://hyperboleandahalf.blogspot.com/2010/04/alot-is-better-than-you-at-everything.html

**Comment by MrPetovan** [comment]
Joke aside, I agree with you, running it on all the incoming images, including the ones that don't have text, is wasteful.

**Comment by shiribailem** [comment]
it's been a really long day, week, month... I'm moving out this weekend to couch surf for probably 3 months before I can get into a new place after a huge blow up with my sibling/roommate... Hopefully you can forgive some sloppy grammer /s /lol

**Comment by shiribailem** [comment]
Anyways, if it's going to be run server side, it should also have tight constraint controls to prevent it from eating all available resources. But honestly I think it's best client side since it can done dynamically (ie. scan just when the image loads on the user end, back-filling the alt-text as part of the javascript cycle. It probably doesn't even need any UI design work no it's most basic implementation as it would just be "scan this and when done put the results here" that triggers when it loads a post with an image.

Though, I would recommend some consideration toward UI design because it would be even better to have a control for it (especially if you run into something with unsatisfactory alt-text, like they included a joke instead of actual transcription)

**Comment by MrPetovan** [comment]
> it's been a really long day, week, month... I'm moving out this weekend to couch surf for probably 3 months before I can get into a new place after a huge blow up with my sibling/roommate... Hopefully you can forgive some sloppy grammer /s /lol

I'm sorry about that, and sharing the Hyperbole and a Half blog post wasn't to mock you, but to spread Allie Brosh's fantastic Alot drawings.

Back to the matter at hands, all your suggestions are good, but it's frontend stuff, and we aren't the most diligent about it.

**Comment by shiribailem** [comment]
it's okay, that's why I put /s (sarcasm) /lol (humor/lighthearted)

Yeah, I know how it is, mostly I make the suggestion the way I did because the basic version that parallels what's being done on the backend requires no UI design work, just a frontend insert.

Fingers crossed because I think this is really helpful but I suspect that backend Tesseract may not be practical without significant resources.

**Comment by annando** [comment]
Tesseract is quite resource hungry. We only execute it on posts with pictures without an image description. But sadly a lot of posts are without any picture description.

**Comment by shiribailem** [comment]
Yeah, I think the initial thing even if keeping the server side implementation is to create some controls on the resource usage. Like limiting how many instances of Tesseract can be running at one time.

But either way I think the problem of how many posts it has to process keeps running into needing some level of frontend support, whether or not you include the client side module as the frontend is the only way I can think of to prioritize images instead of just doing them all in no particular order.
